### PR TITLE
Add OP to Contracts Mapping & Overwrite Existing

### PR DESCRIPTION
**Pull request by MSilb7**
- Based off a fork of https://github.com/duneanalytics/spellbook/pull/4319
- Migrates contracts to a sector spell
- Adds OP Mainnet (`optimism` in Dune) to the contracts mapping spell
- Maintain backwards compatibility for existing optimism contracts tables by making them views coming from the new cross-chain materialized tables.
- Update any downstream models
- Update readmes

**Steps:**
1. `contracts_<blockchain>_base_starting_model`: incremental build, pulls creation traces and joins with transactions to pull extra necessary fields (i.t. tx from, tx to, tx method id) & also handles for contracts created multiple times (via self-destruct) by only pulling the last instance of each contract.
2. `contracts_<blockchain>_base_iterated_creators`: incremental build, checks which contracts we need to re-map creators too (either new contracts, contracts that have a now known deterministic creator in their lineage, contracts with an AA method in their lineage (need to find a way to better filter these over time). Then iterate through a contracts creators (i.e. contract created by contract created by contract...) to create the final creator<>contract mappings. This also checks if a contract is an ERC20 via logs, since we don't yet have a table with all ERC20 tokens per chain, like we do with NFTs.
3. `contracts_<blockchain>_ontract_mapping`: A view that maps metadata (i.e. creator<> project names, manual overrides, token info).
4. `contracts_<blockchain>_contract_mapping` by chain is then combined to a cross-chain view contracts_contract_mapping

### CODE REVIEW

**Comment by MSilb7** [comment]
The `reinit_rank` step in intermediate is the one that takes a while to run. Trying to optimize it for incremental builds.

**Comment by MSilb7** [comment]
omg i fixed the incremental problem (I think / hope)

**Comment by MSilb7** [comment]
Had to do a logic fix, but should be good now.

**Comment by MSilb7** [comment]
- Used a CTE to filter down to the contracts that we actually need to include in the iterations, so the joins and unions are smaller (this helped a TON on the incremental joins)
- ^ After I did that, I found some logic we could simplify since the initial model would rebuild on any updates to the static refs

File: dbt_project.yml [hunk 1]
```diff
@@ -823,12 +823,6 @@ models:
 
     contracts:
       +schema: contracts
```
Comment (jeff-dude) [inline dbt_project.yml:825]: dbt compile is throwing warnings on this still being here, as no objects exist there anymore:
https://github.com/duneanalytics/spellbook/actions/runs/6935445414/job/18865590337?pr=4747#step:8:12

you can remove these two lines too, it seems.

**Review by jeff-dude — CHANGES_REQUESTED** [review]
few minor comments, making really solid progress 🙏

File: models/_sector/contracts/optimism/_schema.yml [hunk 2]
```diff
@@ -1,42 +1,34 @@
...
-          - unique
-          - not_null
-      - name: mapped_name
-        description: "Mapped contract name"
-
-
   - name: contracts_optimism_contract_mapping
     meta:
-      blockchain: optimism
+      blockchain: base
```
[the remark this answers is not in the document]
Comment (MSilb7) [reply models/_sector/contracts/optimism/_schema.yml:6]: good catch, copy and paste error

File: dbt_project.yml (hunk 1 above)
Comment (MSilb7) [reply dbt_project.yml:825]: removed

File: models/op/token_distributions/optimism/op_token_distributions_optimism_project_name_mapping.sql [hunk 3]
```diff
@@ -35,7 +35,7 @@ WITH unified_mappings AS (
 
                 SELECT
                 dune_name AS proposal_name, mapped_name AS project_name
-                FROM {{ ref('contracts_optimism_project_name_mappings') }}
+                FROM {{ ref('contracts_project_name_mappings') }}
```
[the remark this answers is not in the document]
Comment (MSilb7) [reply models/op/token_distributions/optimism/op_token_distributions_optimism_project_name_mapping.sql:38]: actually, for runtime I'll cut this. It was here as a "might as well," but this whole spell should likely by reorganized to push ref links towards the end as well (later on).

**Review by jeff-dude — APPROVED** [review]
i think we're finally there 🚀 

i'll mark this ready to merge, then wait for a good time to do on it's own for easy monitoring in prod for any potential downstream impacts / performance

**Comment by MSilb7** [comment]
> i think we're finally there 🚀
> 
> i'll mark this ready to merge, then wait for a good time to do on it's own for easy monitoring in prod for any potential downstream impacts / performance

Awesome, I want to keep adding more chains (i.e. Zora next), but we can keep those PRs open for a few days post-merge so we can observe.

**Pull request merged**